test/sender.js

```diff
@@ -0,0 +1,169 @@
+        done()
+      })
+    })
+  });
+
+  it('sends object locally', done => {
+    const server = new TestServer(insecureOptions, () => {
+      const sender = senderLib.create({
+        ...insecureOptions,
+        objectMode: true,
+      })
+      sender.on('error', done)
+      sender.write(messageObject)
+      server.waitFor('data', data => {
+        String(data).should.containEql(messageString)
+        String(data).should.containEql('{')
+        String(data).should.containEql('}')
+        String(data).should.containEql('hi')
+        String(data).should.containEql(year)
+        sender.end()
+        server.close()
+        done()
+      })
+    })
+  })
+  it('sends on TLS', done => {
+    const server = new TestServer(serverOptions, () => {
+      const sender = senderLib.create(clientOptions)
+      sender.on('error', done)
+      sender.send(messageString)
+      server.waitFor('data', data => {
+        String(data).should.containEql(messageString)
+        String(data).should.containEql(year)
+        sender.send(messageObject)
+        server.waitFor('data', data => {
+          String(data).should.containEql(messageString)
+          String(data).should.containEql('{')
+          String(data).should.containEql('}')
+          String(data).should.containEql('hi')
+          String(data).should.containEql(year)
+          sender.end()
+          server.close()
+          done()
+        })
+      })
+    })
+  })
+  it('sends insecurely to TLS', done => {
+    const server = new TestServer(serverOptions, () => {
+      const sender = senderLib.create(insecureOptions)
+      for (let i = 0; i < 1000; i++) {
+        sender.send(messageString)
+      }
+      sender.on('error', error => {
+        error.code.should.equal('ECONNRESET')
+        server.close()
+        done()
+      })
+    })
+  })
+  it('sends TLS to insecure', done => {
+    const server = new TestServer(insecureOptions, () => {
+      const sender = senderLib.create(clientOptions)
+      for (let i = 0; i < 1; i++) {
+        sender.send(messageString)
+      }
+      server.waitFor('data', data => {
+        // 0: TLS record type: handshake (22)
+        data[0].should.equal(22)
+        // 1,2: major-minor version, TLS 1.0 is 3,1
+        data[1].should.equal(3)
+        server.close()
+        done()
+      })
+    })
+  })
+})
+
+class TestServer {
+  constructor(options, callback) {
+    const libnet = options.cert ? tls : net
+    this._server = libnet.createServer(options, socket => {
+      this._socket = socket
+      this._socket.on('error', error => this.emit(error))
+    })
+    this._server.on('error', error => console.error('puchi %s', error))
+    this._server.unref()
+    this._server.listen(options.port, callback)
+  }
+
+  waitFor(event, handler) {
+    if (this._socket) {
+      return this._socket.once(event, handler)
+    }
+    setImmediate(() => this.waitFor(event, handler))
+  }
+
+  close() {
+    this._server.close()
+  }
+}
+
```

Comment:
I am missing testing the sender as a stream

Reply:
There are two tests but they were not clearly labeled, changed in commit 282a62a.

Reply:
what I mean by "testing the sender as a stream" is using it like:
```
const sender = devo.sender();
somerReadable.pipe(sender);
```
and check that it send the message(s)

Reply:
That is equivalent to calling to `sender.write()` many times. I am adding a new test that sends many messages and checks backpressure, but it is not yet ready.